apps/blas/dot/dot.fil

```diff
@@ -0,0 +1,217 @@
+  x_1: ['G, 'G+1] W,
+  x_2: ['G, 'G+1] W,
+  x_3: ['G, 'G+1] W,
+  x_4: ['G, 'G+1] W,
+  x_5: ['G, 'G+1] W,
+  x_6: ['G, 'G+1] W,
+  x_7: ['G, 'G+1] W,
+  x_8: ['G, 'G+1] W,
+  x_9: ['G, 'G+1] W,
+  x_10: ['G, 'G+1] W,
+  x_11: ['G, 'G+1] W,
+  x_12: ['G, 'G+1] W,
+  x_13: ['G, 'G+1] W,
+  x_14: ['G, 'G+1] W,
+  x_15: ['G, 'G+1] W,
+  y_0: ['G, 'G+1] W,
+  y_1: ['G, 'G+1] W,
+  y_2: ['G, 'G+1] W,
+  y_3: ['G, 'G+1] W,
+  y_4: ['G, 'G+1] W,
+  y_5: ['G, 'G+1] W,
+  y_6: ['G, 'G+1] W,
+  y_7: ['G, 'G+1] W,
+  y_8: ['G, 'G+1] W,
+  y_9: ['G, 'G+1] W,
+  y_10: ['G, 'G+1] W,
+  y_11: ['G, 'G+1] W,
+  y_12: ['G, 'G+1] W,
+  y_13: ['G, 'G+1] W,
+  y_14: ['G, 'G+1] W,
+  y_15: ['G, 'G+1] W,
+) -> (
+  out: ['G+L, 'G+L+1] W
+) with {
+  let N = 8;
+  let S = 16;
+  let W = 32;
+  let II = S/N;
+  some L where L > 0;
+} where L > 0 {
+
+  Dx := new Dot[W, S, N];
+
+  bundle x[S]: for<k> ['G, 'G+1] W;
+  bundle y[S]: for<k> ['G, 'G+1] W;
+
+  x{0} = x_0; x{1} = x_1; x{2} = x_2; x{3} = x_3; 
+  x{4} = x_4; x{5} = x_5; x{6} = x_6; x{7} = x_7;
+  x{8} = x_8; x{9} = x_9; x{10} = x_10; x{11} = x_11;
+  x{12} = x_12; x{13} = x_13; x{14} = x_14; x{15} = x_15;
+
+  y{0} = y_0; y{1} = y_1; y{2} = y_2; y{3} = y_3; 
+  y{4} = y_4; y{5} = y_5; y{6} = y_6; y{7} = y_7;
+  y{8} = y_8; y{9} = y_9; y{10} = y_10; y{11} = y_11;
+  y{12} = y_12; y{13} = y_13; y{14} = y_14; y{15} = y_15;
+
+  d := Dx<'G>(x{0..S}, y{0..S});
+
+  out = d.out;
+
+  L := Dx::L;
+}
+
+// W: Width of nums
+// S: Size of vectors
+// N: Number of multipliers
+comp Dot[W, S, N]<'G:reuses>(
+  go: interface['G],
+  x[S]: ['G, 'G+1] W,
+  y[S]: ['G, 'G+1] W
+) -> (
+  out: ['G+L, 'G+L+1] W
+) with {
+  some L where L > 0;
+  let reuses = S / N;
+} where W > 0,
+        L > 0,
+        S % N == 0
+{
+
+  M := new Multipliers[W, N];
+  RA := new RAdd[W, N];
+
+  // needed for typechecking
+  assume RA::L == log2(N);
+  assume M::II <= reuses;
+
+  let latency = reuses*M::II+M::L + RA::L;
+  L := latency;
+
+  bundle mult_out[reuses][N]: for<k> ['G+k*M::II+M::L, 'G+k*M::II+M::L+1] W;
+  bundle add_out[reuses]: for<q> ['G + q*M::II + M::L + RA::L, 'G + q*M::II + M::L + RA::L + 1] W;
+  
+  P := new Prev[W, 1];
+  FA := new Add[W];
+
+  for j in 0..reuses {
+
+    let mul_start = j*M::II;
+    let mul_end = j*M::II + M::L;
+    let last_add_start = j*M::II + M::L + RA::L;
+
+    // first use can use raw inputs
+    if j == 0 {
+      m := M<'G+mul_start>(x{j*N..(j+1)*N}, y{j*N..(j+1)*N});
+      mult_out{j}{0..N} = m.out{0..N};
+    // any other use should use registered inputs
+    } else {
+      x_concat := new ConcatBundle[W, N]<'G>(x{j*N..(j+1)*N});
+      y_concat := new ConcatBundle[W, N]<'G>(y{j*N..(j+1)*N});
+      x_concat_reg := new Shift[W*N, mul_start]<'G>(x_concat.out);
+      y_concat_reg := new Shift[W*N, mul_start]<'G>(y_concat.out);
+      x_bundle_reg := new SplitWire[W, N]<'G+mul_start>(x_concat_reg.out);
+      y_bundle_reg := new SplitWire[W, N]<'G+mul_start>(y_concat_reg.out);
```

Comment:
This pattern of concat -> shift -> split seems reusable. We could consider changing the shift register definition to accept bundles instead of values and default it to 1

Reply:
#417